README.rst
Site Auditor
============

Для чего
--------
Этот инструмент необходим для анализа и аудита сайтов. Осуществляет быстрый и качественный сбор информации в автоматическом режиме, достаточно указать цель.

Особенности
-----------
- Open Source
- WHOIS
- ip, title, description, keywords, web-server, powered by, content language, content type
- Яндекс ТИЦ, Google PR, Alexa rank (во всем мире/в отдельной стране)
- Проверка на наличие в каталогах Яндекс, Mail, Yahoo, DMOZ, TDP
- Количество ссылок в Яндекс Блоги (часто требует капчу), Google, Яндекс (сколько проиндексировано всего (часто требует капчу)/попаввшие в индекс), Yahoo, Bing (в индексе, кол-во исходящих ссылок)
- Проверка установки Яндекс метрики, Google Analytics, Live Internet, Rambler TOP100, Mail Rating
- Проверка существования страниц авторизации - Joomla, WordPress, UMI.CMS, Ucoz, Bitrix, /admin, /login, MODX, DLE, Drupal, ISP Manager
- Вывод ``sitemap.xml`` и ``robots.txt``, если существуют.
- W3C HTML validator
- Безопасный просмотр

Пример работы
-------------
.. code-block::

    Enter site, please: google.ru
    Full scan? y/n y
    ==================================================
    WHOIS
    ==================================================
    domain:        GOOGLE.RU
    nserver:       ns1.google.com.
    nserver:       ns2.google.com.
    nserver:       ns3.google.com.
    nserver:       ns4.google.com.
    state:         REGISTERED, DELEGATED, VERIFIED
    org:           Google Inc.
    registrar:     RU-CENTER-REG-RIPN
    admin-contact: https://www.nic.ru/whois
    created:       2004.03.04
    paid-till:     2014.03.05
    free-date:     2014.04.05
    source:        TCI

    nic-hdl:       RU-CENTER-REG-RIPN
    org:           Regional Network Information Center
    phone:         +7 495 737 0601
    fax-no:        +7 495 737 0602
    e-mail:        ru-bill@nic.ru
    www:           https://www.nic.ru/whois
    whois:         whois.nic.ru
    source:        TCI

    Last updated on 2014.02.09 16:31:33 MSK
    ==================================================
    Base site information
    ==================================================
    Site ip - 87.245.196.177
    Web Server - gws
    Powered by - NO
    Content Language - NO
    Content Type - text/html; charset=UTF-8
    Site title - Google
    Description - NO
    Key words - NO
    W3C HTML validator - 26 Errors, 3 warning(s)
    ==================================================
    Ranks
    ==================================================
    Yandex TYC - 64000
    Google Page Rank - 7
    Alexa Rank in all world - 42
    Alexa Rank in Russia - 3
    ==================================================
    Catalogs
    ==================================================
    Yandex Catalog - YES
    Mail Catalog - YES
    Yahoo Catalog - NO
    DMOZ Catalog - YES, 16
    TDP Catalog - YES
    ==================================================
    Links
    ==================================================
    Yandex Blog links - 418563
    Proindexirovano v Google - примерно 1340000
    Proindexirovano v Yandex - 2 млн
    Popavshie v index Yandex - 123875
    Yahoo index - 3,320,000
    Bing index - 3190000
    Bing Outgoing Links - 59600
    ==================================================
    Stats
    ==================================================
    Yandex Metrika - NO
    Google Analytics - NO
    Live Internet - NO
    Rambler TOP100 - NO
    Mail Rating - NO
    ==================================================
    Admins
    ==================================================
    Joomla Admin Directory - NO
    WordPress Admin Directory - NO
    UMI.CMS Admin Directory - NO
    Ucoz Admin Directory - NO
    Bitrix Admin Directory - NO
    Simple Login Page - NO
    Simple Admin Login Page - NO
    MODX Admin Directory or ISP Manager - NO
    DLE Admin Directory - NO
    Drupal Login page - NO
    ==================================================
    Safe Browsing
    ==================================================
    Google - NO - В настоящее время этот сайт не занесен в список подозрительных. NO - За последние 90 дней на этом сайте не размещалось вредоносное ПО.
    Yandex - Сайт google.ru не заражён, либо подробности заражения ещё не опубликованы
    Site Advisor - This link is safe. We tested it and didn't find any significant security issues
    ==================================================
    Files
    ==================================================
    Robots.txt: EMPTY
    SiteMap XML: EMPTY
    ==================================================
    All Time - 14.97 seconds
    ==================================================


Установка
---------

Для работы **Site Auditor** необходим  `Requests <https://github.com/kennethreitz/requests>`_  и Python 3.4.

Вы можете `скачать <https://github.com/stleon/OmgSite/archive/master.zip>`_ текущую версию (все самое новое). Или вы также способны на следующее:

.. code-block::

    pip install requests
    pip install site-auditor


Как пользоваться
----------------

На данный момент мы на стадии **Pre-Alpha**. Вы можете увидеть сообщения об ошибках и т.д. Но **Site Auditor** работает на
большинстве протестированных мной сайтов.

.. code-block::

    cd site_auditor
    python site_auditor.py

Куда будет развиваться
----------------------
Планируем сделать:

- Web-приложение
- Анализ позиций сайтов по запросам
- Статистика по каждому запросу
- Возможность ранжирования запросов
- Система контроля работы сайта с оповещением

Copyright (C) 2014 ST LEON

email: leonst998@gmail.com

web site: http://omgit.ru

HISTORY.rst
История изменений
=================

1.4.1 (22.03.2014)
-----------------
Исправлено:

- определение кол-ва ссылок с яда

1.4.0 (17.03.2014)
-----------------
Изменено:

- Никаких 'YES'or 'NO' в атрибутах, теперь True or False

1.3.8 (17.03.2014)
-----------------
Добавлено:

- Python 3.4

1.3.7 (16.03.2014)
-----------------
Изменено:

- Новый вид сканирования

1.3.6 (10.03.2014)
-----------------
Изменено:

- ``SiteException`` перенесен в ``site_auditor.py``


1.3.5 (09.03.2014)
-----------------
Добавлено:

- багфиксы, адаптация под модели saw

1.3.1 - 1.3.4 (26.02.2014)
-----------------
Добавлено:

- pip install
- новая реализация __str__

Удалены:

- *txt файлы (out)

1.3.0 (26.02.2014)
-----------------
Добавлено:

- 2 вида сканирования - полное и частичное

1.2.6 - 1.2.7 (25.02.2014)
-----------------
Добавлено:

- Docs fix

1.2.5 (19.02.2014)
-----------------
Добавлено:

- Site Advisor

1.2.4 (17.02.2014)
-----------------
Добавлено:

- Вывод времени, затраченного на анализ

1.2.3 (11.02.2014)
-----------------
Добавлено:

- Поиск чистого домена

1.2.2 (11.02.2014)
-----------------
Добавлено:

- Исходящие ссылки Bing
- Страницы в индексе Bing


1.2.1 (11.02.2014)
-----------------
Добавлено:

- TDP Catalog
- Yahoo index

1.2.0 (10.02.2014)
-----------------
Добавлено:

- Безопасный просмотр

1.1.2 (10.02.2014)
-----------------
Удалено:

- Неиспользуемый код

1.1.1 (10.02.2014)
-----------------
Добавлено:

- Новая функция парсера

1.1.0 (09.02.2014)
-----------------
Добавлено:

- ``Microdata`` в тестовом режиме
- Обновлены доки

1.0.5 (09.02.2014)
-----------------
Добавлено:

- Теперь строковое представление объекта можно настраивать в ``out.txt``
- Обновлены доки

1.0.4 (08.02.2014)
-----------------
Исправлено:

- Никаких "!№" и тд в домене

1.0.3 (08.02.2014)
-----------------
Исправлено:

- Домен не менее 4х символов
- ``www`` теперь не удаляются

1.0.2 (07.02.2014)
-----------------
Исправлено:

- Домен не длиннее 255 символов
- Работа с доменами, написанными на кириллице

1.0.1 (07.02.2014)
-----------------
Изменения:

- Нет u'', так как Python 3

1.0.0 (07.02.2014)
-----------------
Изменения:

- Переход на Python 3.3

0.2.1 (05.02.2014)
-----------------
Добавлено:

- W3C HTML validator

0.2.0 (05.02.2014)
-----------------
Добавлено:

- Whois

0.1.0 (04.02.2014)
-----------------
Добавлено:

- Проверка существования страниц авторизации MODX, DLE, Drupal, ISP Manager

0.0.1 (04.02.2014)
-----------------
Добавлено:

- класс ``SiteAuditor`` и класс ``MetaHTMLParser``

site_auditor/meta_parser.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
from html.parser import HTMLParser


class MetaHTMLParser(HTMLParser):
	""" Gets meta description and meta keywords """
	def __init__(self):
		HTMLParser.__init__(self)
		self.meta = {'description': 'NO', 'keywords': 'NO'}

	def handle_starttag(self, tag, attrs):
		"""
		Checks, if meta tags are well written. If true - save it in self.meta
		"""
		if tag.lower() == 'meta':
			#print attrs
			if len(attrs[0]) == 2 and len(attrs[0]) == 2 and attrs[0][1].lower() in self.meta.keys() \
				and attrs[1][0].lower() == 'content':
				self.meta[attrs[0][1].lower()] = attrs[1][1]

if __name__ == '__main__':
	p = MetaHTMLParser()
	p.feed('<html><head><title>Test</title><meta name="description" content="123123"><meta '
								'name="keywords" content="11111"></head>''<body><h1>Parse me!</h1></body></html>')
	print(p.meta)
	p.close()

"""
html = '<html><head><title>Test</title><meta name="description" content="123123"><meta ' \
								'name="keywords" content="11111"></head>''<body><h1>Parse me!</h1></body></html>'

class MetaHTMLParser(HTMLParser):
	def __init__(self):
		HTMLParser.__init__(self)
		self.meta = {'description': 'NO', 'keywords': 'NO'}

	def handle_starttag(self, tag, attrs):
		if tag.lower() == 'meta':

			if len(attrs[0]) == 2 and len(attrs[0]) == 2 and attrs[0][1].lower() in self.meta.keys() \
				and attrs[1][0].lower() == 'content':
				self.meta[attrs[0][1].lower()] = attrs[1][1]


def test1(html):
	p = MetaHTMLParser()
	p.feed(html)
	print(p.meta['description'])
	print(p.meta['keywords'])
	p.close()

# %timeit -n 1000 test1(html)
# 1000 loops, best of 3: 316 µs per loop
# %timeit -n 10000 test1(html)
# 10000 loops, best of 3: 313 µs per loop

def find_meta_tags_content(html, meta_tag):
		meta_tag_content = 'NO'
		match_meta_tag_str = re.compile(r'<meta[.\s\S]+?>', re.I)
		meta_tag_raws = re.findall(match_meta_tag_str, html)
		for raw in meta_tag_raws:
			if meta_tag in raw.lower():
				content_match = re.compile(r'content=(?:\'|\")([\s\S.]+)(?:\'|\")', re.I)
				if re.search(content_match, raw):
					meta_tag_content = re.search(content_match, raw).group(1)
		return meta_tag_content

def test2(html):
	print(find_meta_tags_content(html, 'description'))
	print(find_meta_tags_content(html, 'keywords'))
# %timeit -n 1000 test2(html)
# 1000 loops, best of 3: 101 µs per loop
# %timeit -n 10000 test2(html)
# 10000 loops, best of 3: 103 µs per loop
"""

site_auditor/site_auditor.py
#!/usr/bin/python3
# -*- coding: utf-8 -*-
import socket
from xml.dom.minidom import parseString
import random
import requests
import re
import time


class SiteException(Exception):
	def __init__(self, message):
		self.message = message

	def __str__(self):
		return self.message


class SiteAuditor():
	def __init__(self, site, full_scan):
		try:
			self.start_time = time.time()
			self.full_scan = self.scan_check(full_scan)
			self.site = self.clear_site_name(site)
			self.headers = self.my_headers()
			self.ip = socket.gethostbyname(self.site)
			self.request = requests.get('http://%s' % self.site, headers=self.headers)
		except (socket.gaierror, requests.exceptions.ConnectionError):
			self.error = 'NO HOST AVAILABLE'
		except SiteException as error:
			self.error = error
		else:
			self.error = None
			if self.full_scan == 'yes':
				self.info_green()
				self.info_red()
			elif self.full_scan == 'no':
				self.info_green()
			else:
				self.info_red()
		finally:
			self.all_time = "%.2f seconds" % (time.time() - self.start_time)

	@staticmethod
	def scan_check(string):
		string = string.lower().strip()
		if string == 'y':
			return 'yes'
		elif string == 'n':
			return 'no'
		elif string == '2w':
			return '2w'
		else:
			raise SiteException('Недопустимый вид сканирования!')

	def info_green(self):
		self.yad = self.yandex()
		self.pr = self.page_rank()
		self.alexa = self.alexa_rank()
		self.pro_index = self.index()
		self.bing = list(map(self.bing_information, ['linkfromdomain:', 'site:']))
		self.safe_site = self.safebrowsing()

	def info_red(self):
		self.whois = self.who()
		self.web_server = self.inf_from_headers('server')
		self.powered_by = self.inf_from_headers('x-powered-by')
		self.content_lanuage = self.inf_from_headers('content-language')
		self.content_type = self.inf_from_headers('content-type')
		self.html = self.request.text
		self.description = self.find_meta_tags_content('description')
		self.key_words = self.find_meta_tags_content('keywords')
		self.title = self.site_title()
		self.ya_metrica = self.analytics('yaCounter')
		self.google_an = self.analytics('google-analytics.com/ga.js')
		self.live_inet = self.analytics('counter.yadro.ru/hit')
		self.rambler_top = self.analytics('counter.rambler.ru/top100.jcn?')
		self.mail_rating = self.analytics('top.list.ru/counter?id=')
		self.dmoz = self.dmoz_rank()
		self.robots_txt = self.robots('robots.txt')
		self.sitemap_xml = self.robots('sitemap.xml')
		self.mail_catalog = self.catalogs('Не найдено', 'http://search.list.mail.ru/?q=')
		self.yahoo_catalog = self.catalogs('We did not find', 'http://dir.search.yahoo.com/search?ei=UTF-8&h=c&p=')
		self.tdp_catalog = self.catalogs('No listings have been found', 'http://www.trustdirectory.org/search.php?what=')
		self.joomla = self.engine('administrator')
		self.word_press = self.engine('wp-login.php')
		self.umi = self.engine('admin/content/sitetree')
		self.ucoz = self.engine('panel')
		self.bitrix = self.engine('bitrix/admin')
		self.simple_login = self.engine('login')
		self.admin_login = self.engine('admin')
		self.modx = self.engine('manager')
		self.dle = self.engine('admin.php')
		self.drupal = self.engine('user')
		self.html_validator = self.html_valid()


	@staticmethod
	def clear_site_name(site):

		def check_symbols(i):
			if i not in '`~!№@"$;%^?&*()+{[]}|\\:<>,\' ':
				return i
			else:
				raise SiteException('Недопустимые символы в домене!')
		# регулярное выражение для поиска чистого домена
		clean_url = re.compile(r'(?:http[s]?)?:?(?:)?(?:/){0,2}([\w-]*.[\w-]*)')
		if re.search(clean_url, site).group(1):
			site = re.search(clean_url, site).group(1)
		site = site.strip()
		if len(site) > 255 or len(site) < 4:
			raise SiteException('Длина домена не должна превышать 255 символов или быть меньше 4!')
		if re.search(r'[а-яА-Я]', site):
			site = site.encode('idna').decode('utf-8')
		return ''.join(list(map(check_symbols, site)))

	def my_headers(self):
		useragents = [
			'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.9.1.3) Gecko/20090913 Firefox/3.5.3',
			'Mozilla/5.0 (Windows; U; Windows NT 6.1; en; rv:1.9.1.3) Gecko/20090824 Firefox/3.5.3 \ '
			'(.NET CLR 3.5.30729)',
			'Mozilla/5.0 (Windows; U; Windows NT 5.2; en-US; rv:1.9.1.3) Gecko/20090824 Firefox/3.5.3 \ '
			'(.NET CLR 3.5.30729)',
			'Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US; rv:1.9.1.1) Gecko/20090718 Firefox/3.5.1',
			'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US) AppleWebKit/532.1 (KHTML, like Gecko) \ '
			'Chrome/4.0.219.6 Safari/532.1',
			'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; WOW64; Trident/4.0; SLCC2; \ '
			'.NET CLR 2.0.50727; InfoPath.2)',
			'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.0; Trident/4.0; SLCC1; .NET CLR 2.0.50727; \ '
			'.NET CLR 1.1.4322; .NET CLR 3.5.30729; .NET CLR 3.0.30729)',
			'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.2; Win64; x64; Trident/4.0)',
			'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0; SV1; .NET CLR 2.0.50727; InfoPath.2)',
			'Mozilla/5.0 (Windows; U; MSIE 7.0; Windows NT 6.0; en-US)',
			'Mozilla/4.0 (compatible; MSIE 6.1; Windows XP)',
			'Opera/9.80 (Windows NT 5.2; U; ru) Presto/2.5.22 Version/10.51',
			# new
			'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/30.0.1599.101 Safari/537.36',
			'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) \ '
			'Chrome/32.0.1700.102 Safari/537.36',
			'Mozilla/5.0 (Linux; Android 4.2.2; ru-ru; SAMSUNG GT-19500 Build/JDQ39) AppleWebKit/535.19 \ '
			'(KHTML, like Gecko) Version/1.0 Chrome/18.0.1025.308 Mobile Safari/535.19',

		]
		refers = [
			'https://www.google.com/?q=',
			'https://www.google.ru/?q=',
			'http://yandex.ru/yandsearch?text=',
			'http://go.mail.ru/search?&q=',
			'http://nova.rambler.ru/search?query='
		]
		return {
			'User-Agent': random.choice(useragents),
			'Cache-Control': random.choice(['no-cache', 'cache']),
			'Accept-Charset': 'ISO-8859-1,utf-8;q=0.7,*;q=0.7',
			'Referer': '%s%s' % (random.choice(refers), self.site),
			'Keep-Alive': random.randint(110, 120),
			'Connection': 'keep-alive',
		}

	def inf_from_headers(self, string):
		try:
			return self.request.headers[string]
		except KeyError:
			return False

	def site_title(self):
		try:
			return self.html.split('<title>')[1].split('</title>')[0].strip()
		except IndexError:
			return False

	def analytics(self, string):
		return True if string in self.html else False

	def yandex(self):
		r = requests.get('http://bar-navig.yandex.ru/u?ver=2&url=http://%s&show=1' % self.site)
		xmldoc = parseString(r.text.encode('windows-1251'))
		tyc = xmldoc.getElementsByTagName('tcy')[0].attributes['value'].value
		if tyc == '-1':
			tyc = 0
		# TODO Yandex needs capcha
		r = requests.get('http://blogs.yandex.ru/search.xml?link=%s&noreask=1' % self.site, headers=self.headers).text
		try:
			blog_links = r.split(' из <b>')[1].split('</b> найденных.')[0].strip()
		except IndexError:
			blog_links = 'NEED CAPTCHA'
		#r = requests.get('http://yandex.ru/yandsearch?text="*.%s"&noreask=1&lr=193' % (self.site),).text
		#r = r.split('Нашлось<br>')[1].split('ответов</strong>')[0].strip()
		#print r
		return {
				'tyc': tyc, 'blogs': blog_links,
				'cat': True if xmldoc.getElementsByTagName('textinfo')[0].childNodes[0].nodeValue.strip() else False,
		}

	def dmoz_rank(self):
		try:
			r = requests.get('http://www.dmoz.org/search?q=%s' % self.site, headers=self.headers).text
			return 'YES, %s' % (r.split('Open Directory Sites</strong>')[1].split(')</small>')[0].split(' of ')[1])
		except IndexError:
			return False

	def alexa_rank(self):
		r = requests.get('http://data.alexa.com/data?cli=10&dat=snbamz&url=%s' % self.site, headers=self.headers)
		xmldoc = parseString(r.text.encode('utf-8'))
		try:
			alexa = {
				'all_world': xmldoc.getElementsByTagName('POPULARITY')[0].attributes['TEXT'].value,
				'country': xmldoc.getElementsByTagName('COUNTRY')[-1].attributes['NAME'].value,
				'rank_in_country': xmldoc.getElementsByTagName('COUNTRY')[-1].attributes['RANK'].value,
			}
		except IndexError:
			alexa = {'all_world': 0, 'country': None, 'rank_in_country': None}
		return alexa

	def page_rank(self):
		"""
		#   Desc    :   Get PageRank of a Website, for Python
		#   Author  :   iSayme
		#   E-Mail  :   isaymeorg@gmail.com
		#   Website :   http://www.isayme.org
		#   Date    :   2012-08-04
		https://github.com/isayme/google_pr/blob/master/GooglePr.py
		"""
		gpr_hash_seed = "Mining PageRank is AGAINST GOOGLE'S TERMS OF SERVICE. Yes, I'm talking to you, scammer."
		magic = 0x1020345
		for i in iter(range(len(self.site))):
			magic ^= ord(gpr_hash_seed[i % len(gpr_hash_seed)]) ^ ord(self.site[i])
			magic = (magic >> 23 | magic << 9) & 0xFFFFFFFF
		url = "http://toolbarqueries.google.com/tbr?client=navclient-auto&features=Rank&ch=%s&q=info:%s" % \
				("8%08x" % (magic), self.site)
		data = requests.get(url, headers=self.headers).text
		data = data.split(":") if data else None
		return int(data[len(data) - 1]) if data else 0

	def robots(self, data):
		r = requests.get("http://%s/%s" % (self.site, data), allow_redirects=False)
		return False if r.status_code in [404, 301, 302] or r.text == self.html or 'не найдена' in r.text else r.text

	def index(self):
		google = requests.get("https://www.google.ru/search?q=site:%s" % self.site, headers=self.headers).text
		if 'ничего не найдено' in google:
			google = 0
		else:
			#.replace('примерно', '')
			try:
				google = google.split('id="resultStats">Результатов: ')[1].split('<')[0].replace('&#160;', '')
				if 'примерно' in google:
					google = google.split()[1]
				google = int(google)
			except IndexError:
				google = 'CANT DETECT'
		yad_session = requests.Session()
		yandex = yad_session.get("http://yandex.ru/yandsearch?text=&site=%s&ras=1" % self.site, headers=self.headers)
		try:
			yandex = yandex.text.split('Нашлось<br>')[1].split('ответов</strong>')[0].replace('&nbsp;', ' ').strip()
			if 'млн' in yandex:
				yandex = yandex.split()[0]+'000000'
			if 'тыс' in yandex:
				yandex = yandex.split()[0]+'000'
			yandex = int(yandex)
		except IndexError:
			# TODO Yandex needs capcha
			yandex = 'NEED CAPTCHA'
		yad_ind = requests.get("http://webmaster.yandex.ru/check.xml?hostname=%s" % self.site, headers=self.headers)
		mirror_text = 'Сайт является зеркалом</span> '
		if mirror_text in yad_ind.text:
			mirror = yad_ind.text.split(mirror_text)[1].split(', ')[0]
			yad_ind = requests.get("http://webmaster.yandex.ru/check.xml?hostname=%s" % mirror, headers=self.headers)
		try:
			yad_ind = yad_ind.text.split('Страницы: ')[1].split('</div>')[0]
		except IndexError:
			yad_ind = 0
				# TODO DELETE Retries
		try:
			yahoo = requests.get('http://search.yahoo.com/search?ei=UTF-8&p=site:%s' % self.site,
							 headers=self.headers).text.split('<span>')[-1].split(' result')[0]
			if ',' in yahoo:
				yahoo = yahoo.replace(',','')
			yahoo = int(yahoo)
		except IndexError:
			yahoo = 0
		return {'google': google, 'yandex_standart': yandex, 'yandex_in_index': yad_ind, 'yahoo_index': yahoo}

	def bing_information(self, string):
		try:
			return requests.get('http://www.bing.com/search?q=%s%s' % (string, self.site),
						headers=self.headers).text.split('результаты: ')[1].split('</span>')[0].replace('&#160;', '')

		except IndexError:
			return 0

	def catalogs(self, string, catalog):
		return False if string in requests.get("%s%s" % (catalog, self.site), headers=self.headers).text else True

	def engine(self, string):
		r = requests.get('http://%s/%s' % (self.site, string), headers=self.headers, allow_redirects=True)

		def check(string, text=r.text):
			return True if string not in text else False

		return True if r.status_code not in [404, 403, 503, 301, 302] \
				and r.text != self.html and check('404') and check('не найдена') and check('not found') else False

	def who(self):
		# Не юзается "встроенная" whois, тк на unix она встроенная, на винде, говорят, нет
		r = requests.post("http://www.ripn.net/nic/whois/whois.cgi", {'Whois': self.site, 'Host': 'whois.ripn.net'},
			headers=self.headers, allow_redirects=True,).text.split('(in English).\n')[1].split('</PRE>')[0].strip()
		# http://www.ripn.net/about/servpol.html
		if 'You have exceeded allowed connection rate.' in r:
			return 'Вы сделали больше 30 запросов в минуту (лимит). Попробуйте позже'
		elif 'You are not allowed to connect' in r:
			return 'Вы в течении 15 минут превышали лимит (30 запросов в минуту). Восстановление доступа будет ' \
										'не менее, чем через час'
		else:
			return re.compile(r'<.*?>').sub('', r)

	def html_valid(self):
		r = requests.get('http://validator.w3.org/check?uri=%s' % self.site, headers=self.headers,).text
		r = r.split('valid">')[2].split('</td>')[0].strip()
		return re.compile(r'<.*?>').sub('', r) if '<' in r else r

	def find_meta_tags_content(self, meta_tag):
		"""
		Функция ищет в html коде сайта мета тег переданный ей и возвращает
		его значение. Функция учитывает возможность расположения параметра name
		после параметра content и до.
		
		Принимает:
			meta_tag - значение какого мета тега ищем - title, keywords, descriptions
		
		Возвращает:
			meta_tag_content - строка с содержанием мета-тега
		"""
		meta_tag_content = False
		match_meta_tag_str = re.compile(r'<meta[.\s\S]+?>', re.I)
		meta_tag_raws = re.findall(match_meta_tag_str, self.html)
		for raw in meta_tag_raws:
			if meta_tag in raw.lower():
				content_match = re.compile(r'content=(?:\'|\")([\s\S.]+)(?:\'|\")', re.I)
				if re.search(content_match, raw):
					meta_tag_content = re.search(content_match, raw).group(1)
		return meta_tag_content

	def __str__(self):
		return 'Instance of %s, adress %s:\n%s' % (self.__class__.__name__, id(self), self.__attrnames())

	def __attrnames(self):
		result = ''
		for attr in sorted(self.__dict__):
			result += '\tname %s = %s\n' % (attr, self.__dict__[attr])
		return result

	def safebrowsing(self):
		# This site is not currently listed as suspicious.
		# this site has not hosted malicious software over the past 90 days
		g = requests.get('http://safebrowsing.clients.google.com/safebrowsing/diagnostic?site=%s' % self.site,
							headers=self.headers).text
		if 'is not currently listed':
			message = 'NO - В настоящее время этот сайт не занесен в список подозрительных.'
		else:
			message = 'YES - В настоящее время этот сайт занесен в список подозрительных.'  # may be, need test site
		if 'has not hosted malicious' in g:
			message = '%s NO - За последние 90 дней на этом сайте не размещалось вредоносное ПО.' % message
		else:
			message = '%s YES - За последние 90 дней на этом сайте размещалось вредоносное ПО.' % message
		yad = requests.get('http://yandex.ru/infected?l10n=ru&url=%s' % self.site,
								headers=self.headers).text.split('<title>')[1].split('</title>')[0].strip()
		site_advisor = requests.get('http://www.siteadvisor.com/sites/%s' % self.site,
								headers=self.headers).text.split('class="intro">')[1].split('.</p>')[0]
		return {'google': message, 'yandex': yad, 's_a': site_advisor}


if __name__ == '__main__':
	print(SiteAuditor(input('Enter site, please: '), input('Full scan? y/n/2w ')))
